Add ESLint for TypeScript greatness

### .eslintrc.js
```js
module.exports = {
  parser: '@typescript-eslint/parser',
  extends: [
    'plugin:@typescript-eslint/recommended'
  ],
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'module',
  },
  root: true,
  rules: {
    '@typescript-eslint/array-type': ['error', { default: 'generic' }],
  },
}
```

### package.json
```diff
   "devDependencies": {
     "@types/node": "^14.0.3",
     "@types/node-fetch": "^2.5.7",
+    "@typescript-eslint/eslint-plugin": "^3.0.2",
+    "@typescript-eslint/parser": "^3.0.2",
     "ajv": "^6.12.2",
+    "eslint": "^7.1.0",
     "node-fetch": "^2.6.0",
     "npm-run-all": "^4.1.5",
     "ts-json-schema-generator": "^0.68.0",
     "ts-node": "^8.10.1",
     "typescript": "^3.9.3",
     "typescript-json-schema": "^0.42.0"
   },
```

---

## OUTPUT

```sh
npm run lint

> rapid-experiments-shared@0.0.1 lint /Volumes/Dev/github/mozilla/rapid-experiments-shared
> eslint .


/Volumes/Dev/github/mozilla/rapid-experiments-shared/bin/build-schemas.ts
  26:9  warning  'program' is assigned a value but never used   @typescript-eslint/no-unused-vars
  83:9  error    'rv' is never reassigned. Use 'const' instead  prefer-const

/Volumes/Dev/github/mozilla/rapid-experiments-shared/test/normandy-arguments.ts
  51:43  warning  'promise' is defined but never used  @typescript-eslint/no-unused-vars

/Volumes/Dev/github/mozilla/rapid-experiments-shared/types/messaging.ts
  3:12  error  Don't use `object` as a type. The `object` type is currently hard to use ([see this issue](https://github.com/microsoft/TypeScript/issues/21732)).
Consider using `Record<string, unknown>` instead, as it allows you to more easily inspect and use the keys  @typescript-eslint/ban-types

/Volumes/Dev/github/mozilla/rapid-experiments-shared/types/normandy.ts
  112:27  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 5 problems (2 errors, 3 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -12,6 +12,7 @@ jobs:
             - run: yarn install
             - run: yarn build
             - run: yarn test
+            - run: yarn lint
       - run:
           name: Preparing artifacts
           command: yarn artifact

--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -1,0 +1,17 @@
+module.exports = {
+  parser: "@typescript-eslint/parser",
+  extends: [
+    "plugin:@typescript-eslint/recommended"
+  ],
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: "module"
+  },
+  root: true,
+  rules: {
+    "@typescript-eslint/array-type": [
+      "error",
+      { default: "generic" }
+    ]
+  }
+};

--- a/bin/build-schemas.ts
+++ b/bin/build-schemas.ts
@@ -80,7 +80,7 @@ async function* walk(
   opts: WalkOptions = { includeDirs: "exclude" }
 ): AsyncGenerator<WalkDirEntry, void> {
   for await (const entry of await fs.opendir(dir)) {
-    let rv = entry as WalkDirEntry;
+    const rv = entry as WalkDirEntry;
     rv.path = path.join(dir, entry.name);
     if (entry.isDirectory()) {
       yield* walk(rv.path, opts);

--- a/package.json
+++ b/package.json
@@ -1,10 +1,14 @@
 {
   "name": "rapid-experiments-shared",
-  "description": "This is a place to define data and schemas used across the rapid experiments program.",
   "version": "0.0.1",
-  "author": "Mozilla",
-  "bugs": {
-    "url": "https://github.com/mozilla/rapid-experiments-shared/issues"
+  "license": "MPL-2.0",
+  "scripts": {
+    "build": "npm-run-all -s build:*",
+    "build:json-schema": "./bin/build-schemas.ts",
+    "artifact": "./bin/pack-artifact.sh",
+    "lint": "eslint .",
+    "test": "npm-run-all -s test:*",
+    "test:normandy-arguments": "./test/normandy-arguments.ts"
   },
   "dependencies": {},
   "devDependencies": {
@@ -21,20 +25,5 @@
     "typescript": "^3.9.3",
     "typescript-json-schema": "^0.42.0"
   },
-  "homepage": "https://github.com/mozilla/rapid-experiments-shared#readme",
-  "keywords": [],
-  "license": "MPL-2.0",
-  "private": true,
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/mozilla/rapid-experiments-shared.git"
-  },
-  "scripts": {
-    "artifact": "./bin/pack-artifact.sh",
-    "build": "npm-run-all -s build:*",
-    "build:json-schema": "./bin/build-schemas.ts",
-    "lint": "eslint .",
-    "test": "npm-run-all -s test:*",
-    "test:normandy-arguments": "./test/normandy-arguments.ts"
-  }
+  "private": true
 }

--- a/package.json
+++ b/package.json
@@ -1,24 +1,40 @@
 {
   "name": "rapid-experiments-shared",
+  "description": "This is a place to define data and schemas used across the rapid experiments program.",
   "version": "0.0.1",
-  "license": "MPL-2.0",
-  "scripts": {
-    "build": "npm-run-all -s build:*",
-    "build:json-schema": "./bin/build-schemas.ts",
-    "artifact": "./bin/pack-artifact.sh",
-    "test": "npm-run-all -s test:*",
-    "test:normandy-arguments": "./test/normandy-arguments.ts"
+  "author": "Mozilla",
+  "bugs": {
+    "url": "https://github.com/mozilla/rapid-experiments-shared/issues"
   },
   "dependencies": {},
   "devDependencies": {
     "@types/node": "^14.0.3",
     "@types/node-fetch": "^2.5.7",
+    "@typescript-eslint/eslint-plugin": "^3.0.2",
+    "@typescript-eslint/parser": "^3.0.2",
     "ajv": "^6.12.2",
+    "eslint": "^7.1.0",
     "node-fetch": "^2.6.0",
     "npm-run-all": "^4.1.5",
     "ts-json-schema-generator": "^0.68.0",
     "ts-node": "^8.10.1",
     "typescript": "^3.9.3",
     "typescript-json-schema": "^0.42.0"
+  },
+  "homepage": "https://github.com/mozilla/rapid-experiments-shared#readme",
+  "keywords": [],
+  "license": "MPL-2.0",
+  "private": true,
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/mozilla/rapid-experiments-shared.git"
+  },
+  "scripts": {
+    "artifact": "./bin/pack-artifact.sh",
+    "build": "npm-run-all -s build:*",
+    "build:json-schema": "./bin/build-schemas.ts",
+    "lint": "eslint .",
+    "test": "npm-run-all -s test:*",
+    "test:normandy-arguments": "./test/normandy-arguments.ts"
   }
 }

--- a/types/messaging.ts
+++ b/types/messaging.ts
@@ -1,4 +1,5 @@
 export interface CfrMessage {
   slug: string;
+  /* eslint-disable-next-line @typescript-eslint/ban-types */
   trigger: object;
 }

